### DISCUSSION TITLE
Allow marking BPP APIs to be sync as part of network policy

### DISCUSSION THREAD

**Discussion by pkapustin** [beckn/protocol-specifications#187] [Ideas]
# Background

Currently, we often see this repeating pattern in our BPP implementations when processing a request:
* Start a background task
* Return ACK
* Perform a callback to the BAP in a background task

In addition to doing extra work, there are multiple potential points of failure here, and implementing proper error handling becomes a complex task.

# Proposal
We suggest to simplify this where possible by allowing to mark certain BPP APIs as synchronous for HTTP-based implementations as part of the network policy. 
If a BPP API is marked as sync, instead of returning ACK, the response body would contain the data that would otherwise come in a callback.

# Example
A confirm BPP API is marked as sync in a taxi network. A customer confirms a taxi ride order. A BAP issues confirm API, and the BPP uses the sync response to communicate that the order has been placed with all the necessary information. No implementation of on_confirm is required. 

# Summary
We see that allowing to mark BPP APIs to be sync as part of network policy gives the following benefits:
- Simpler implementation
- More robust code
- Better performance

**Comment by ravi-prakash-v** [comment]
I agree, this is a valid issue. And I also agree that the network policy _can_ choose to make the message exchange mechanism synchronous. In such a case, as long as the request body has `context.action` equal to say, `search` and the (synchronous) response body has the `context.action` equal to `on_search`, along with an error object, the acknowledgement can be _implied_ or even _concatenated_ with the message in the synchronous response. While it is technically possible, we need to also think about the practical implications of this change. Not that these are unsolvable problems but we need to acknowledge them and provide recommendations and solutions to address the same. Below are some implications that I see. 

**1. Documentation**: The transaction protocol needs to be _decoupled_ from the transport layer protocol. That means, the documentation of the specification can no longer be in Open API 3.x format as it assumes HTTP by default. Maybe Async API would make more sense. 

**2. Interoperability** :  If two networks are adopting two different exchange mechanisms, how will we ensure interoperability between them. Do we need separate bridging interfaces or can this be baked into the protocol spec itself?

**3. Migration Impact** : Moving from an async model to a synchronous model may introduce tech debt in existing implementations. Although if this mechanism does become a standard, it will be easier to do this sooner than later when the adoption becomes large. 

Adding @pramodkvarma to comment on the same.

**Comment by pkapustin** [comment]
Yes, all these are valid points of course.

Just wanted to share a couple of possible (and pragmatic) solutions to these.
1. **Documentation**. Yes, if core spec wants to be agnostic of transport protocol, OpenAPI is not ideal. However, I think that OpenAPI can still be a good choice for documenting a network-specific instance of the core spec, if the network is HTTP-based. If such network chooses to implement some BPP APIs synchronously, this can also be clearly expressed using OpenAPI.
2. **Interoperability**. I think it is important to note that two different networks will use two different sets of policies documented by their network-specific specs, and that means that usually they are not going to be automatically interoperable. Marking a BPP API as sync is just one example of a network policy. However, if there is a need for two different networks to work together, we can consider bridging solutions, etc.
3. **Migration impact**. Agree, I think it would be nice to recommend considering this in a new network. But it's not like all existing networks need to change to this model (and it wouldn't be feasible in all cases either). I see it more of a useful simplification for relatively new networks.

**Comment by pkapustin** [comment]
We have successfully implemented this as a pilot in KOMN taxi network.

**Reply by ravi-prakash-v** [reply]
Currently, core-0.9.3 doesn't support sync transactions. A new NP might implement async and try to plug in the network. However this is a mission-specific issue and can be moved out of this thread.

**Comment by ravi-prakash-v** [comment]
An important point that I would like to highlight is **scalability**. This was missed out in my previous comment. Sync APIs don't scale well in high volume transactions. Since it is an open network, the volume of transactions as new NPs (especially BAPs) get added scale up steeply. To manage such spike loads on the network, I believe async is the way to go. 

Also async also allows for deferred confirmations like "driver accepting a ride", "seller manually checking his inventory and confirming an order" and so on. Not all booking platforms provide confirmations immediately. Some booking platforms confirm an order on a queue basis. 

Adding @venkatramanm @pramodkvarma @rahul101001000 to comment on the same.

**Reply by pkapustin** [reply]
I agree that scalability is important. However, scalability is not black and white, and so isn't sync / async. For example, most popular languages support calling sync web APIs in an async manner. So for the developer it looks like they are getting a synchronous response, leading to much simpler code and error handling, but internally it is implemented as a callback. This way, no threads are blocked at all, but the developer is spared from a lot of the complexities.
I think most Google APIs are normal web APIs with requests and responses, and that scales pretty well? 

Web APIs are very popular and work well. Forcing callbacks everywhere will inevitably mean that there will be a lot of missed callbacks. We see that a lot in our implementations. Of course, the BPP can implement retries, but if the BPP hasn't received a response from the first callback (suppose a connection was closed), can it be sure that no handler code was executed at the BAP side? So now the callback API needs to be idempotent, and this may be tricky to implement. 
These problems go away with normal sync web APIs. Also the number of calls is twice as little that also helps with scalability.

I think we need to remember that many of the Beckn users and adopters are not necessarily experts in distributed computing. Let's give them a choice, and if they prefer to use simple tools, they can do it, and let their languages and platforms handle the low-level issues like making sure no threads are blocked, etc.

In addition, there is another aspect to it. If an API is marked as sync in a network spec, this means that it has to return reasonably quickly. This creates clear expectations, now the BPP is responsible for implementing that API in a responsive way. There may be more work to do for the BPP, but that can be put on a queue, and BAP can receive a sync response using a simple interface. If it is async, there are no guarantees, the BPP may reply in several seconds. This will negatively impact the user experience. We need responsive systems right?

However, I do agree completely that in many cases one would want to use async APIs. One typical use case is to communicate events originating at the BPP side 
https://github.com/beckn/protocol-specifications/discussions/177
Actually, in this case it is not async API, rather a normal web API on the BAP side that is called by the BPP when an event happens.

There will also be use cases like @core-wg-admin mentioned, for example when it may indeed take a lot of time to confirm an order. Sure, in this case such API may be kept async, there is nothing wrong with that.

To summarize, I think there are merits to both approaches. And while personally I do think that usually a combination of normal sync web APIs and an API to communicate events would often be preferred, I don't want to impose this view on anyone, I think it is important that the implementors have the options. If we would like people to use Beckn protocol, let's give them some freedom, so it helps them to build their products, otherwise they wouldn't want to do it.

Sometimes you need a normal web api, sometimes you need a callback pair, sometimes you need an API for events. I suggest we don't make it a choice between a hammer and a screwdriver. All of these have their merits in different situations.

**Reply by ravi-prakash-v** [reply]
You're right. Google APIs are sync and scale well. But we're talking about a protocol that is agnostic of form or scale. A small store owner with am edge server on say, a raspberry pi or a cell phone might not have the compute capability to handle thousands of requests. And that goes for cloud applications as well. Many cloud based BPPs may also not have the resources to scale up. Sync works for large platforms with high parallel processing capability but we cannot have one protocol for platforms that scale well and another for platforms that do not.

**Reply by ravi-prakash-v** [reply]
Secondly, if it is a matter of developer convenience, we can always create SDKs and libraries that make the job easier for a developer.

**Comment by pramodkvarma** [comment]
Btw, in reality, many commerce transactions, while get ack in sync mode, don't necessarily complete in sync mode. Assuming everything must happen in sync mode fundamentally defeats the long lived nature of real world transactions. For example, I should be able to book a hotel, make payment 1 hr later, hotel confirms booking end of day, and actually fulfill, say, a month later when I stay or cancel it few days before the date, etc. It's important that all functional steps are unbundled and async to allow such long living transactions to happen. 

And in many networks, calls are cascaded to other systems before actual confirmation can happen. For example, when I select a cab option in a taxi booking app it has to connect to various near by drivers, wait for one of the drivers to accept, and then confirm back. While consumer is waiting and experiences in sync fashion, no system is actually blocking anything, behind the scene these steps are all fulfilled in async fashion. 

Of course, if every platform uses async methods to separate backend and cascading calls and then synchronously respond, then this is somewhat manageable. But, in reality this assumes every app/platform on the network has implemented this which never happens. So, designing these functional steps as explicit async call back APIs, in one way, we are enforcing the async logic building and scale across platform builders on beckn enabled network.

**Comment by rahul101001000** [comment]
While we are all used to `sync APIs` with say [Amazon Seller Partner API (SP-API)](https://github.com/amzn/selling-partner-api-docs/blob/main/references/catalog-items-api/catalogItemsV0.md) or [Shopify API](https://shopify.dev/api/admin-rest), the reason why they can conform to stringent SLAs is because all catalog, order, ... data is stored with them. And these companies have the expertise to meet these sync API SLAs. 

In the case of Beckn, it is fundamentally an unbundled system as @pramodkvarma said. We have to account for it not being feasible for BPPs to meet real time SLAs. Hence `async APIs` for Beckn.

Now, a few improvements can be made to Beckn - 
1. As @core-wg-admin suggested, conforming to [AsyncAPI](https://github.com/asyncapi/spec/blob/v2.2.0/spec/asyncapi.md) instead of OpenAPI may be better suited to Beckn. That also moves us closer to an [events driven paradigm](https://www.asyncapi.com/docs/getting-started/coming-from-openapi) than being message driven, which I personally prefer more. That said, I'm not sure if moving to events driven from message driven is a need / goal for the Beckn design at this point. 

2. Alternative idea - have we explored using [HTTP 202 Accepted](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202) with the Beckn spec specifying that the header return an async-session-id if a real time response cannot be provided. The async-session-id can be then used to poll for status. See  [Facebook Graph API](https://developers.facebook.com/docs/graph-api/reference/async-session) as an example. Using this, the base API can be `sync`, but with the option of turning `async` with a 202.

**Comment by pramodkvarma** [comment]
Thanks @rahul101001000. Btw, beckn ack response is equivalent to HTTP 202 Accepted. Basically server is responding to the API with a functional ack (works even when HTTP is not used) to bring the same effect. But, instead of polling from client side, we preferred a push (call back from API server to API client) since both end points are trusted platforms on the network (unlike browser and server).

**Reply by venkatramanm** [reply]
Would like to add one more point.  Since polling typically puts load on both systems, push is more preferred.

**Comment by rahul101001000** [comment]
> Btw, beckn ack response is equivalent to HTTP 202 Accepted. 

Aah good to know, thanks for the info @pramodkvarma . 

Any thoughts on adopting AsyncAPI and/or being more event friendly?

**Reply by venkatramanm** [reply]
@rahul101001000 , I did see a  problem statement in innovation category for the ONDC hackathon  on this topic of Async Api.

**Reply by rahul101001000** [reply]
Thanks @venkatramanm !  Oh yes, seeing it now... great to see AsyncApi adoption on the cards.

**Comment by venkatramanm** [comment]
@pramodkvarma has mentioned all the key points I wanted to share. My take is all Apis should be designed and built for Async. Sync requirements for a bap by usecase can be absorbed possibly into "a" gateway design. 

BAPs that need some things  to be synchronous can make that api call with "this" kind of a gateway. 
which can make it look synchronous and respond with a consolidated response.   

To address @pkapustin 's concern of complexity, I think early adopters could come up with robust open source sdks that can be consumed by later adopters. This way the complexity can be contained and addressed at an sdk level.

**Comment by pkapustin** [comment]
For the sanity of this already complex discussion, I suggest to leave the topic of moving core schema from OpenAPI to AsyncAPI (or other formats) out, let's have a separate discussion for that.

**Comment by pkapustin** [comment]
@core-wg-admin 

>Sync works for large platforms with high parallel processing capability

Can we expand this a bit, why are you saying web APIs don't work for small platforms? What exactly is the problem? I am still not quite seeing how a response like "Thank you for your order, we'll update you when a courier is on their way" scales worse than "ACK" response followed by immediate callback saying the same? There are twice as many calls in the second case?

@core-wg-admin
>But we're talking about a protocol that is agnostic of form or scale.

>we cannot have one protocol for platforms that scale well and another for platforms that do not.

Suppose we have a network with a BAP that generates thousands of orders per minute, and we have a BPP running on a cell phone where an owner needs to check the inventory manually every time. This is not going to work whether they have sync or async APIs, right?
That's why I think network implementors need to have a working group where they discuss these things. And if they agree that in their context it makes sense to mark certain APIs as sync, because it helps with anything (scalability, responsiveness, simplicity, error rates, etc), I don't think we can or should try to prevent that? I think you also agreed earlier in this thread that this could be decided as part of the network policy? Also, agree we should provide guidelines and standards, but I think these can differ for various spec adaptations, and take transport protocol into account (see also https://github.com/beckn/protocol-specifications/issues/186).

@core-wg-admin @venkatramanm  
> Secondly, if it is a matter of developer convenience, we can always create SDKs and libraries that make the job easier for a developer.

> To address @pkapustin's concern of complexity, I think early adopters could come up with robust open source sdks that can be consumed by later adopters. This way the complexity can be contained and addressed at an sdk level.

Ths has already been solved by language and library implementors, we can just use it?
This code doesn't block any threads on the client side, response is handled in a callback, and the error handling complexity is taken care of by the platform.
```
async function callAPI() {
  try {
    const res = await axios.get(...);
    // ...
  } catch(error) {
    //...
  }
}
```

@pramodkvarma 
> Btw, in reality, many commerce transactions, while get ack in sync mode, don't necessarily complete in sync mode. Assuming everything must happen in sync mode fundamentally defeats the long lived nature of real world transactions. For example, I should be able to book a hotel, make payment 1 hr later, hotel confirms booking end of day, and actually fulfill, say, a month later when I stay or cancel it few days before the date, etc. It's important that all functional steps are unbundled and async to allow such long living transactions to happen.

Of course, I agree and I didn't mean that everything has to complete synchronously. In reality it's often not just request / reply either, we often have multiple events happening at the BPP over time and have to be communicated back to the BAP. I think also @rahul101001000 and @venkatramanm suggested it may be a good idea to start thinking more in terms of events.
https://github.com/beckn/protocol-specifications/discussions/177

Let's consider some examples:
1. Customer confirms a trip, gets an order id. Later events from BPP: driver assigned, driver cancelled, trip reassigned, trip started, trip ended.
2. Customer buys a ticket, gets a confirmation. Later events from BPP: ticket available for download, ticket scanned, departure delayed, departure cancelled.
3. BAP orders a delivery, BPP confirms the order. Later events from BPP: courier out for delivery, package picked up, package delivered.
4. Customer cancels their taxi trip, BPP confirms that the trip will be cancelled. Later events from BPP: trip cancelled.
5. Customer updates their trip, changing the route, BPP confirms the change. Later events from BPP: new driver allocated (with new pickup time).

In our experience such situations are easier modeled and implemented as one sync request followed by a series of BPP events rather than a request / callback pair followed by a series of BPP events.

Let's consider example 5. We can do either:
1. Update with a sync response from BPP. On_event from BPP later.
2. Update with an ACK from BPP followed by immediate on_update from BPP. On_event from BPP later. 

Note: currently we don't have on_event, so for now one needs to use on_update, but I think on_event is much cleaner for this purpose (see also https://github.com/beckn/protocol-specifications/discussions/177).

We think that for HTTP implementations, the first alternative is much easier, better, faster, less error prone to implement. Consider just the validation part: what if the order the customer is trying to update an order that does not exist? With a sync response they will get a validation error immediately. With request / callback pair, what should happen? ACK as long as the schema is valid, followed by immediate on_update saying your order does not exist? Everything gets much more complex, especially error handling, uses more time and processing power. In addition, in our experience, the code at BAPs to handle errors coming as requests from BPP very rarely gets written at all. The same applies to the error handling code at BPP that is supposed to handle errors during the callback request to the BAP.

What I am essentially proposing is "extended ACK" that allows for more information to be included with the ACK. Everything is still unbundled, because the real work at the BPP side will happen later anyway, but the code is simpler and faster.

@pramodkvarma 
> Of course, if every platform uses async methods to separate backend and cascading calls and then synchronously respond, then this is somewhat manageable. But, in reality this assumes every app/platform on the network has implemented this which never happens. So, designing these functional steps as explicit async call back APIs, in one way, we are enforcing the async logic building and scale across platform builders on beckn enabled network.

I agree that for implementations using persistent message queues for communication, this would be a natural style to communicate. But for HTTP implementations, I think things are different. Sure, enforcing that the only thing that can be returned in a web API is ACK has an advantage that there is a lower chance that this request part of the API will be slow. But there are many disadvantages to it as well (more verbose, complexity related to general implementation, error handling, retries, idempotence, validation, higher chance for errors, more calls leading to more computations, responsiveness issues as it is not clear when the callback should come). 

Simplicity is a very important factor here. Not every user of Beckn is an expert in implementing distributed systems. Let's embrace existing tools. We have async / await in most languages, and that let's you write simple and correct code that updates an order, handling the situation when the order does not exist (or any other error) by using a normal try / catch, and without any threads being blocked on the client side. I think it is better to use what is availalbe and not try to go against the wind.

I think we can still implement good guidelines and help implementors to avoid common pitfalls, and we can come up with more useful guidelines and standards, depending on the transport protocol used by the specific adaptation. 

But as I tried to show, this is clearly a trade-off, and in case of HTTP implementation, I think it is not in favor of request / callback pairs. That's why I don't think we should (or have the authority to) to impose a rule like "no one is allowed to return anything except ACK in any API in any network".

**Comment by pramodkvarma** [comment]
@pkapustin You are making an interesting argument. I am OK for us to explore an elegant way to allow this and still have seamless interoperability across versions and providers. I am not in favor of networks deciding these as policies, ideally this should be baked into protocol level approach as @rahul101001000 pointed out in the case of HTTP. 

Few thoughts:
1. Who decides the implementation in sync or async? I suppose the API provider should decide (BPP in the above example). We should continue to keep `search` async, right? I presume we are talking about post search calls.
2. How does BAP detect if BPP has implemented sync or async? Do we need some meta APIs or by declaring it in the network participant registry? I prefer latter to avoid more APIs. Let BPP declare (at version level what their implementation is and BAPs can cache these).
3. What will BPP return in the case of sync implementation? Entire response as if they would have done in say `on_select` call back? I would like to keep it simple and not end up creating more schemas or partial schemas. Either BPP return full response in sync mode or async call back mode. BAP should have same logic of processing the responding either way instead of them having to deal with two different responses.
4. Do we need any enhancement for `on_update` like a generic `on_event` as @pkapustin suggested?
5. If we go this route, then minimal backward compatibility issues (BAP API calling logic will need to handle this to allow BPPs to implement sync mode).

@core-wg-admin @venkatramanm @rahul101001000 what do you guys think?

@pkapustin while this is technically possible and I am aligned with your arguments, should we do this now? Given UPI transactions are all async (there also we had this argument in the first year, now everyone takes it for granted) and with tech folks easily implementing it, I am still unsure if we should complicate this. Wouldn't tech/sdk providers easily abstract this?

**Reply by ChakshuGautam** [reply]
@pramodkvarma @core-wg-admin @rahul101001000 What would be a good abstraction for the SDK? These options come to mind.
- A generic async to sync converter outside of core specification built on `asyncapi` spec so that all the tooling built by `asyncapi` can be used
- at a use case level implemented by individual providers

**Comment by pramodkvarma** [comment]
Thanks @ChakshuGautam. Of course, people may want similar utils in other languages as well or as nginx plugin etc. Could we attempt this? If this works smoothly, then we don't need to change the core beckn design at all!

Would such a util service work the following way? (I am trying to better understand myself)
1. Intercept the call (say, `select` API on BPP side)
2. Start an async call to API internally
3. Send Ack to caller (BAP)
4. Wait for the internal API response
5. Construct `on_select` and call back

I am wondering if such intercepting async to sync proxy util is feasible? How does failures (at each stage) get handled? If this works, surely worth exploring!

**Reply by ChakshuGautam** [reply]
Yes, was thinking to implement exactly as you mentioned.

> How does failures (at each stage) get handled?

Currently, I am thinking of aggressively raising errors when they happen on both `x` and `on_x` API pairs. I think I can cook up a POC quickly and share. The end goal would be to implement something like [asyncapi generators](https://github.com/asyncapi/generator) that allows us to generate this using the base spec assuming no deviation in implementation.

**Comment by pkapustin** [comment]
@ChakshuGautam @pramodkvarma
While the SDKs and converters are possible, this sounds like a lot of work that has to be done for all implementing platforms, also, this will add runtime overhead? Also, this is quite error-prone as well? And most of this is already implemented in async / await?

In addition I don't understand why we are talking about changing the design of the core spec?
As @core-wg-admin mentioned, core spec is agnostic of transport protocol. So the fact that sometimes we want to send a reply as a response for HTTP implementations is only natural, right?

Here is what I suggest:

- Convert core spec to any format of preference
- Create adaptations of core spec for major transport protocols, this is anyway needed, also see https://github.com/beckn/protocol-specifications/issues/186

Then we don't need to change the core spec, and it should solve everything without any implementation or runtime overhead, right?

Now, if we would like to bridge two networks, say, HTTP-based and web sockets-based, such SDKs/converters will be necessary, but then their implementation will be fully justified.

**Comment by pkapustin** [comment]
@pramodkvarma 

>  I am not in favor of networks deciding these as policies, ideally this should be baked into protocol level approach as @rahul101001000 pointed out in the case of HTTP

Agree, this can be part of the HTTP adaptation of the core spec and maintained by Beckn.

> Who decides the implementation in sync or async? I suppose the API provider should decide (BPP in the above example)

Makes sense, but can also be discussed in the network working group, as all BPPs in that network will have to comply for interoperability.

> We should continue to keep search async, right? I presume we are talking about post search calls.

Correct.

>How does BAP detect if BPP has implemented sync or async? Do we need some meta APIs or by declaring it in the network participant registry? I prefer latter to avoid more APIs. Let BPP declare (at version level what their implementation is and BAPs can cache these).

Yes, we can do something like this. Also, like with many other policies, the implementations don't really support dynamic policies yet, as this significantly complicates the implementation. So while in general dynamic policies are desired, I think we could start with making it a static network policy decided by the network working group.

> What will BPP return in the case of sync implementation? Entire response as if they would have done in say on_select call back? I would like to keep it simple and not end up creating more schemas or partial schemas. 

Correct.

> Either BPP return full response in sync mode or async call back mode. BAP should have same logic of processing the responding either 
way instead of them having to deal with two different responses.

Agree, but as mentioned earlier I think it can be made a static policy in the beginning as it simplifies the implementation and helps achieve mission objectives.

> Do we need any enhancement for on_update like a generic on_event as @pkapustin suggested?

Not exactly related to this discussion, but I strongly recommend it. There has to be a way for a BPP to communicate fine-grained changes to the BAP, on_update doesn't currently support that, and events are a great fit.

> @pkapustin while this is technically possible and I am aligned with your arguments, should we do this now? 

I would recommend a small pilot to get more experience and evaluate this. We have actually implemented confirm and cancel as sync APIs in KOMN taxi as this helped us to reach some mission objectives. Maybe we can use it to learn more about it and then discuss it again?

> Wouldn't tech/sdk providers easily abstract this?

I think we should really try to do it without creating more work / implementation, runtime overhead, etc. as this kind of defeats the  purpose. I think people can use their async / await, that will be faster, simpler, better integrated, with fewer errors, and no work required.
Also, we don't need to change the core spec at all, this can be part of the HTTP adaptation.

What do you think?

**Comment by pramodkvarma** [comment]
Quick comment. I do not prefer networks making static decision. Frankly not all BPPs need to implement the same way. Let whichever BPP need to implement sync implement it and simply declare their API version as sync. Any large network shouldn't assume everyone will have same version, same methods, etc. If we are going down this road, we should be flexible architecturally to allow BPPs to simply switch from sync to async or vice versa mid flight while the network is live. 

But, before we go down the path, let us do both experiments. One suggested by @ChakshuGautam will mean no change and what you are suggesting will mean BAP changes to handle both sync and async responses depending on BPP implementation. Let us do a spike and see what it tells us, study the results, and discuss.

Given national networks are on its way using current spec, anyway, we need to be careful changing this in a hurry. Hope that works.

**Comment by ravi-prakash-v** [comment]
Breaking down the above thread into the following sub-points that require consensus
1. Are we in agreement that a declaration of sync / async is needed?
2. Can we publish the next version of the protocol in Async API format instead of Open API format?
3. Can we add provision for declaring sync / async implementation in the next version of the protocol i.e 0.9.4? (Estimated Release Date : Jan 31st, 2022)
4. Can we take the SDK approach to handle the async behaviour first until sync declaration gets supported in the protocol?

Expecting a yes/no from the participants so that we can proceed. I have added my comments below

**Reply by ravi-prakash-v** [reply]
> Are we in agreement that a declaration of sync / async is needed?

Yes

> Can we publish the next version of the protocol in Async API format instead of Open API format?

Yes

> Can we add provision for declaring sync / async implementation in the next version of the protocol i.e 0.9.4? (Estimated Release Date : Jan 31st, 2022)

No, not until a consensus is achieved. Maybe in 0.9.5. I feel it is too soon to make that change.

> Can we take the SDK approach to handle the async behaviour first until sync declaration gets supported in the protocol?

Yes

**Reply by pkapustin** [reply]
>Are we in agreement that a declaration of sync / async is needed?

Yes, for the HTTP adaptation.

>Are we publishing the next version of the protocol in Async API format instead of Open API format?

I don't really have an opinion on this yet. Though, I don't mind doing it as long as we release the Open API-based HTTP adaptation at the same time.

>Are we adding provision for declaring sync / async implementation in the next version of the protocol i.e 0.9.4? (Estimated Release Date : Jan 31st, 2022)

I think we need to discuss a bit more on the network policy aspect of this to decide.

>Are we taking the SDK approach to handle the async behaviour first until sync declaration gets supported in the protocol?

At this point I am not sure SDK approach is a good solution. I think we can live with async until sync is supported in the HTTP adaptation of the spec.

**Reply by venkatramanm** [reply]
> Breaking down the above thread into the following sub-points that require consensus
> 
> 1. Are we in agreement that a declaration of sync / async is needed?
Yes and we will need in both sides, BPP can declare by api whether  its going to return "sync", BAP need to declare by api if it can take "sync" response. If a BAP is expecting the response to come with an on_x... call, but BPP says it will send as "sync" . We have an issue. It needs both sides to handle sync and async.There fore to keep things simple, we need to absorb this into the gw. It should see if a bap is sync and respond synchronously to that platform. If a bpp declares it is sync, should take synchronously from the bpp and send as sync or async as needed by the BAP That way, the complexity is absorbed only at the  gw. But yes, registry should have entry by each endpoint being sync or async for BAP and BPP platform. 

> 2. Can we publish the next version of the protocol in Async API format instead of Open API format?
Can 

> 3. Can we add provision for declaring sync / async implementation in the next version of the protocol i.e 0.9.4? (Estimated Release Date : Jan 31st, 2022)
Too soon to add. It think we need to debate on the spec and the registry spec. Need to make mods in implementations of GW. I also think more domains are gettting onboarded at this point. Changing the spec and not having these supports in GW and registry will cause unwarrented disruptions. We can put in the proposal and take it into appropriate version draft. 

> 4. Can we take the SDK approach to handle the async behaviour first until sync declaration gets supported in the protocol?
Yes
> 
> Expecting a yes/no from the participants so that we can proceed. I have added my comments below

**Comment by pramodkvarma** [comment]
1. Not yet in core protocol if proxy pattern works
2. Yes if you all have time to take up
3. Not yet
4. Yes

**Reply by ChakshuGautam** [reply]
Hey, @core-wg-admin is there a discussion I can follow (outside of this one) for the need to move towards `AsyncAPI` rather than _OpenAPI_. Since`AsyncAPI` is a wider and newer spec it may not have that much adoption compared to OpenAPI and adding it as a prerequisite will lead to far too many different implementations. 

I am summarizing the points *for* adoption to AsyncAPI spec below for my own understanding. Please let me know if these are exhaustive. Thanks.
- 1. Existing spec that is natively supporting async paradigm
- 2. Transport protocol agnostic
- 3. Transport message type agnostic

**Comment by rahul101001000** [comment]
Moving to `AyncAPI` also means that we'll be embracing a publish subscribe model compatible with amqp, http, kafka, mqtt, ... - right?. We need to plan for updates to the reference implementation and some documentation on this as well. Just publishing the spec in `AsyncAPI'`, while necessary, won't be sufficient. This would be additional work but worth it - the events paradigm suits Beckn well. @core-wg-admin @pramodkvarma 

Nice suggestion @ChakshuGautam of doing this externally - a lot of sync async requirements are indeed taken care of via libraries.

**Reply by pramodkvarma** [reply]
Yes. But, for now http continues to be the preferred choice for most.

**Reply by pkapustin** [reply]
I think it is important to acknowledge this and create the HTTP adaptation so people don't make these choices in each network.
See also https://github.com/beckn/protocol-specifications/issues/186

We can later create other adaptations for pub / sub, etc.

**Reply by pramodkvarma** [reply]
Yes

**Reply by pkapustin** [reply]
@pramodkvarma If you agree, could you comment on the issue?
https://github.com/beckn/protocol-specifications/issues/186

**Comment by ChakshuGautam** [comment]
Hey @pramodkvarma @core-wg-admin @rahul101001000 
Based on the discussion earlier I had a discussion with [Anshul](https://github.com/AnshulMalik) and he has [created a POC ](https://github.com/AnshulMalik/reference-sandbox/tree/sync-proxy) on the same which handles all `x` and `on_x` pairs in a generic way using Nodejs. This is a quick and dirty solution for a quick feedback. He has used the `reference-sandbox` to create a [sync_proxy](https://github.com/AnshulMalik/reference-sandbox/tree/sync-proxy/sync-proxy).

Please review. This will also help us narrow down on what a proxy should and should not do. Thanks.

**Reply by pramodkvarma** [reply]
Wow! That was fast @ChakshuGautam and @AnshulMalik! Thanks. Is there a flow diagram that helps me better understand? Is it implementing the same flow I mentioned once or something else. Did this use actual beckn APIs and schemas?

@core-wg-admin @pkapustin @venkatramanm is there a way we can try this with some real beckn flows in mobility or commerce sandboxes? I am trying to figure out how best to test this approach out. Any suggestions to help Chakshu and Anshul?

**Reply by rahul101001000** [reply]
Nice work @ChakshuGautam and @AnshulMalik! What may also help is a readme.md in the sync proxy folder that shows how this API call from the sync proxy / library will be different than the current calls in the reference-sandbox.

**Reply by ChakshuGautam** [reply]
>  Is there a flow diagram that helps me better understand?

I will share the docs by EOD tomorrow.

> Is it implementing the same flow I mentioned once or something else

Yes

> Did this use actual beckn APIs and schemas?

Yes. Using the implementation on the [reference sandbox](https://github.com/beckn/reference-sandbox) on the beckn org.

**Reply by pramodkvarma** [reply]
Super!

**Reply by venkatramanm** [reply]
> Wow! That was fast @ChakshuGautam and @AnshulMalik! Thanks. Is there a flow diagram that helps me better understand? Is it implementing the same flow I mentioned once or something else. Did this use actual beckn APIs and schemas?
> 
> @core-wg-admin @pkapustin @venkatramanm is there a way we can try this with some real beckn flows in mobility or commerce sandboxes? I am trying to figure out how best to test this approach out. Any suggestions to help Chakshu and Anshul?
In the beckn-one sandbox we can test it out to see request and response jsons. But it may not work with real transactions with parties as every one has coded to the spec which demands a callback. We should probably take on the registry & GW change in a dev branch and experiment if sync and async parties can talk to each other.

**Reply by pkapustin** [reply]
@pramodkvarma With this solution, if I understand it correctly, we first abstract away from web APIs, implementing message-style communication on top of those (core spec), and then create one more extra layer to get the original web API semantics back (proxy). 

I think this is simply not worth it. The point of my proposal is to leverage web APIs and existing tools in HTTP implementations, simplify the code, improve performance, reduce error rates. With this approach, I don't understand how it meets any of those goals, so I don't see how this can be tested and evaluated.

Having said that, I think something like this could be useful if at some point one needs to bridge existing networks that communicate differently. For communication in new networks, I think we need a different approach.

**Comment by pkapustin** [comment]
@venkatramanm
> If a BAP is expecting the response to come with an on_x... call, but BPP says it will send as "sync" . We have an issue. It needs both sides to handle sync and async. 

I think this is a typical example when making a certain network policy dynamic makes the implementation more complex. 
If both BAPs and BPPs have to support both sync and async, this increases the complexity, while the intention of this proposal is to decrease it.

Also, while I agree that in the future we may try to solve this and some other policies it in a dynamic way, the truth is that today almost all policies are solved statically in the networks, and the majority of them are not really feasible to solve in a dynamic way, at least short-term. Like, in one network one field is required, in other it isn't, etc. 

We are currently working on achieving interoperability between participants that need to interoperate (one technical network). 
I think it is okay to accept that achieving it between different technical networks is often not going to be feasible because of all the differences. If such interoperability becomes necessary at some point, we can bridge them.

Basically, even if we put a lot of time and effort into making sync / async in a dynamic policy, I don't think it is going to give us that much, as we are still only going to have interoperability inside a technical network because of all of the other differences. That's why I think that at this point of network development and focus on mission objectives, quick network growth and easily including new participants, I think it is okay to have sync / async as a static policy. We can however have some recommendations at the HTTP spec adaptation level to minimize the differences between various networks. 

> There fore to keep things simple, we need to absorb this into the gw. It should see if a bap is sync and respond synchronously to that platform. If a bpp declares it is sync, should take synchronously from the bpp and send as sync or async as needed by the BAP That way, the complexity is absorbed only at the gw. But yes, registry should have entry by each endpoint being sync or async for BAP and BPP platform.

Implementing it in the gateway does sound interesting. However, I think it has a number of significant drawbacks. This would make the gateway mandatory for all APIs, affecting performance and scalability. Callbacks have to go through the same gateway as requests. In general, I think this would again significantly increase the overall complexity without any real benefits.
Also, as long as we have registry, I think it is better to treat gateway as a good-to-have entity rather than a must-have entity, this simplifies some other things, improves performance and makes the network more flexible.
https://github.com/beckn/protocol-specifications/discussions/169

My recommendations:

- We can leave with async for a little while
- We can define a reasonable set of standards / recommendations for using sync in the HTTP adaptation, including what kinds of APIs should be sync and what APIs should be async. For example, search / on_search is async, short commands are sync, events come through on_update, etc. 

I think this would simplify things and give us some real benefits in terms of performance, error rates, etc.
